pkg/container-collection/container-collection.go

```diff
@@ -0,0 +1,207 @@
+		c := value.(pb.ContainerDefinition)
+		if namespace == c.Namespace && pod == c.Podname && container == c.Name {
+			pid = c.Pid
+			// container found, stop iterating
+			return false
+		}
+		return true
+	})
+	return
+}
+
+// LookupPIDByPod returns the PID of all containers belonging to
+// the pod specified in arguments, indexed by the name of the
+// containers or an empty map if not found
+func (cc *ContainerCollection) LookupPIDByPod(namespace, pod string) map[string]uint32 {
+	ret := make(map[string]uint32)
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(pb.ContainerDefinition)
+		if namespace == c.Namespace && pod == c.Podname {
+			ret[c.Name] = c.Pid
+		}
+		return true
+	})
+	return ret
+}
+
+// GetContainersBySelector returns a slice of containers that match
+// the selector or an empty slice if there are not matches
+func (cc *ContainerCollection) GetContainersBySelector(
+	containerSelector *pb.ContainerSelector,
+) []pb.ContainerDefinition {
+	selectedContainers := []pb.ContainerDefinition{}
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(pb.ContainerDefinition)
+		if ContainerSelectorMatches(containerSelector, &c) {
+			selectedContainers = append(selectedContainers, c)
+		}
+		return true
+	})
+	return selectedContainers
+}
+
+// ContainerLen returns how many containers are stored in the collection.
+func (cc *ContainerCollection) ContainerLen() (count int) {
+	cc.containers.Range(func(key, value interface{}) bool {
+		count++
+		return true
+	})
+	return
+}
+
+// ContainerRange iterates over the containers of the collection and calls the
+// callback function for each of them.
+func (cc *ContainerCollection) ContainerRange(f func(pb.ContainerDefinition)) {
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(pb.ContainerDefinition)
+		f(c)
+		return true
+	})
+}
+
+// ContainerRangeWithSelector iterates over the containers of the collection
+// and calls the callback function for each of those that matches the container
+// selector.
+func (cc *ContainerCollection) ContainerRangeWithSelector(
+	containerSelector *pb.ContainerSelector,
+	f func(pb.ContainerDefinition),
+) {
+	cc.containers.Range(func(key, value interface{}) bool {
+		c := value.(pb.ContainerDefinition)
+		if ContainerSelectorMatches(containerSelector, &c) {
+			f(c)
+		}
+		return true
+	})
+}
+
+// Subscribe returns the list of existing containers and registers a callback
+// for notifications about additions and deletions of containers
+func (cc *ContainerCollection) Subscribe(key interface{}, selector pb.ContainerSelector, f pubsub.FuncNotify) []pb.ContainerDefinition {
+	cc.pubsub.Subscribe(key, func(event pubsub.PubSubEvent) {
+		if ContainerSelectorMatches(&selector, &event.Container) {
+			f(event)
+		}
+	})
+	ret := []pb.ContainerDefinition{}
+	cc.ContainerRangeWithSelector(&selector, func(c pb.ContainerDefinition) {
+		ret = append(ret, c)
+	})
+	return ret
```

Comment:
It could happen that callback is called and the container is returned in the list, right? (if a new container is added after calling `Subscribe` but before `ContainerRangeWithSelector`. I suppose the caller should handle that case, right?

Reply:
I think you are right, since I removed the mutex here...
I think it would be better if the caller does not have to handle this case. I'll think about it.

Reply:
I updated Subscribe() so this race condition should not happen anymore. PTAL.